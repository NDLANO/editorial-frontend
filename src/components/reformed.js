/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * A modified version of https://github.com/davezuko/react-reformed (MIT license)
 */

import React from 'react';
import PropTypes from 'prop-types';
import hoistNonReactStatics from 'hoist-non-react-statics';
import { getComponentName } from 'ndla-util';

const makeWrapper = WrappedComponent => {
  class FormWrapper extends React.Component {
    constructor(props, ctx) {
      super(props, ctx);
      this.state = {
        submitted: false,
        model: props.initialModel || {},
        fields: {},
      };
      this.setModel = this.setModel.bind(this);
      this.setProperty = this.setProperty.bind(this);
      this.bindToChangeEvent = this.bindToChangeEvent.bind(this);
      this.bindInput = this.bindInput.bind(this);
      this.setSubmitted = this.setSubmitted.bind(this);
    }

    componentWillReceiveProps(nextProps) {
      const currentModel = this.props.initialModel;
      const nextModel = nextProps.initialModel;

      if (this.props.resetOnInitialModelChange && currentModel !== nextModel) {
        const hasChanges = Object.keys(nextModel).find(
          key => nextModel[key] !== currentModel[key],
        );
        if (hasChanges) {
          this.setModel(nextModel);
        }
      }
    }

    setModel(model) {
      this.setState({ model });
      return model;
    }

    setSubmitted(submitted) {
      this.setState({ submitted });
    }

    setProperty(prop, value, isDirty = true) {
      const model = {
        ...this.state.model,
        [prop]: value,
      };

      if (this.state.fields[prop] && this.state.fields[prop].isDirty) {
        this.setState(prevstate => ({
          model: { ...prevstate.model, [prop]: value },
        }));
      } else {
        this.setState(prevstate => {
          const fields = { ...prevstate.fields, [prop]: { isDirty } };
          return { model: { ...prevstate.model, [prop]: value }, fields };
        });
      }
      return model;
    }

    bindToChangeEvent(e) {
      const { name, type, value } = e.target;
      if (type === 'checkbox') {
        const oldCheckboxValue = this.state.model[name] || [];
        const newCheckboxValue = e.target.checked
          ? oldCheckboxValue.concat(value)
          : oldCheckboxValue.filter(v => v !== value);

        this.setProperty(name, newCheckboxValue);
      } else if (type === 'EditorState') {
        this.setProperty(name, value, value.getCurrentContent().hasText()); // Only set dirty flag if text has changed
      } else if (type === 'SlateEditorState') {
        this.setProperty(name, value); // TODO: Handle dirty flag with SlateEditorState
      } else {
        this.setProperty(name, value);
      }
    }

    bindInput(name) {
      return {
        name,
        value: this.state.model[name] || '',
        onChange: this.bindToChangeEvent,
      };
    }

    render() {
      const nextProps = {
        ...this.props,
        bindInput: this.bindInput,
        bindToChangeEvent: this.bindToChangeEvent,
        model: this.state.model,
        fields: this.state.fields,
        submitted: this.state.submitted,
        setProperty: this.setProperty,
        setSubmitted: this.setSubmitted,
        setModel: this.setModel,
      };

      return React.createElement(WrappedComponent, nextProps);
    }
  }

  FormWrapper.propTypes = {
    initialModel: PropTypes.object, //eslint-disable-line
    resetOnInitialModelChange: PropTypes.bool.isRequired,
  };
  FormWrapper.defaultProps = {
    resetOnInitialModelChange: false,
  };
  FormWrapper.displayName = `Reformed(${getComponentName(WrappedComponent)})`;
  return hoistNonReactStatics(FormWrapper, WrappedComponent);
};

export default makeWrapper;
