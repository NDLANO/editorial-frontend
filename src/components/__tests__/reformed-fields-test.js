/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import TestRenderer from 'react-test-renderer';
import get from 'lodash/fp/get';
import reformed from '../reformed';

const findInput = (tree, name) =>
  tree.children.find(child => child.props.name === name);

const triggerOnChange = (tree, name, value) => {
  const input = findInput(tree, name);
  input.props.onChange({ type: 'change', target: { value, name } });
};

const triggerEvent = (tree, name, type) => {
  const input = findInput(tree, name);
  input.props.onFocus({ type, target: { name } });
};

export const FieldForm = props => {
  const { bindInput, fields } = props;

  return (
    <form>
      <span {...get('title', fields)} />
      <input type="text" {...bindInput('title')} />
      <span {...get('content', fields)} />
      <input type="text" {...bindInput('content')} />
      <span {...get('image.alt', fields)} />
      <input type="text" {...bindInput('image.alt')} />
      <span {...get('image.caption', fields)} />
      <input type="text" {...bindInput('image.caption')} />
    </form>
  );
};

FieldForm.propTypes = {
  fields: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  bindInput: PropTypes.func.isRequired,
};

const initialModel = {
  title: 'Olé',
  content: 'Olé, Olé, Olé',
  image: {
    caption: '',
    alt: 'Alternative image text',
  },
};

test('reformed HOC sets dirty flag in fields', () => {
  const Reformed = reformed(FieldForm);

  const component = TestRenderer.create(
    <Reformed initialModel={initialModel} />,
  );
  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();

  triggerOnChange(tree, 'title', 'Hom');
  triggerOnChange(tree, 'title', 'Hombre');
  triggerOnChange(tree, 'image.caption', 'El Grande');

  expect(component.toJSON()).toMatchSnapshot();
});

test('reformed HOC sets active flag on focus', () => {
  const Reformed = reformed(FieldForm);

  const component = TestRenderer.create(
    <Reformed initialModel={initialModel} />,
  );
  const tree = component.toJSON();

  triggerEvent(tree, 'title', 'focus');
  triggerEvent(tree, 'image.caption', 'focus');

  expect(component.toJSON()).toMatchSnapshot();
});

test('reformed HOC removes active flag and set touched flag on blur', () => {
  const Reformed = reformed(FieldForm);

  const component = TestRenderer.create(
    <Reformed initialModel={initialModel} />,
  );
  const tree = component.toJSON();

  triggerEvent(tree, 'title', 'focus');
  triggerEvent(tree, 'title', 'blur');
  triggerEvent(tree, 'image.caption', 'focus');
  triggerEvent(tree, 'image.caption', 'blur');

  expect(component.toJSON()).toMatchSnapshot();
});

test('reformed HOC merges and overrides input flags correctly', () => {
  const Reformed = reformed(FieldForm);

  const component = TestRenderer.create(
    <Reformed initialModel={initialModel} />,
  );
  const tree = component.toJSON();

  triggerEvent(tree, 'title', 'focus');
  expect(component.toJSON()).toMatchSnapshot();
  triggerOnChange(tree, 'title', 'test');
  expect(component.toJSON()).toMatchSnapshot();
  triggerEvent(tree, 'title', 'blur');
  expect(component.toJSON()).toMatchSnapshot();
});
