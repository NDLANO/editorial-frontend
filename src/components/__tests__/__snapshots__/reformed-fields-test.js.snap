// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reformed HOC removes active flag and set touched flag on blur 1`] = `
<form>
  <span
    active={false}
    isDirty={false}
    touched={true}
  />
  <input
    name="title"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value="Olé"
  />
  <span />
  <input
    name="content"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value="Olé, Olé, Olé"
  />
  <span />
  <input
    name="image.alt"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value="Alternative image text"
  />
  <span
    active={false}
    isDirty={false}
    touched={true}
  />
  <input
    name="image.caption"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value=""
  />
</form>
`;

exports[`reformed HOC sets active flag on focus 1`] = `
<form>
  <span
    active={true}
    isDirty={false}
  />
  <input
    name="title"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value="Olé"
  />
  <span />
  <input
    name="content"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value="Olé, Olé, Olé"
  />
  <span />
  <input
    name="image.alt"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value="Alternative image text"
  />
  <span
    active={true}
    isDirty={false}
  />
  <input
    name="image.caption"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value=""
  />
</form>
`;

exports[`reformed HOC sets dirty flag in fields 1`] = `
<form>
  <span />
  <input
    name="title"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value="Olé"
  />
  <span />
  <input
    name="content"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value="Olé, Olé, Olé"
  />
  <span />
  <input
    name="image.alt"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value="Alternative image text"
  />
  <span />
  <input
    name="image.caption"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value=""
  />
</form>
`;

exports[`reformed HOC sets dirty flag in fields 2`] = `
<form>
  <span
    isDirty={true}
  />
  <input
    name="title"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value="Hombre"
  />
  <span />
  <input
    name="content"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value="Olé, Olé, Olé"
  />
  <span />
  <input
    name="image.alt"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value="Alternative image text"
  />
  <span
    isDirty={true}
  />
  <input
    name="image.caption"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value="El Grande"
  />
</form>
`;
